feat(addDocuments): addDocuments method and test

Itzaprado opened the pull request:
[ ] Add addEmbeddings method
[ ] Add addTexts method
[ ] Update vectorStore test
[ ] Add customZip as util function

Itzaprado committed:
feat(addDocuments): addDocuments method and test
[ ] Add addEmbeddings method
[ ] Add addTexts method
[ ] Update vectorStore test
[ ] Add customZip as util function

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -0,0 +1,315 @@
+      metadatas = Array.from(texts).map(() => new Object());
+    }
+
+    const tuples = customZip(ids, texts, embeddings, metadatas)
```
I need to review this closer, but I think you need to be using the pgvector-node library https://github.com/pgvector/pgvector-node to handle the vectors appropriately.

Itzaprado committed:
refactor: change addEmbeddings to addVectors

Itzaprado committed:
fix: change content to be pageContent

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -178,15 +180,86 @@ class PostgresVectorStore extends VectorStore {
     )
   }
 
+  async addVectors(vectors: number[][], documents: DocumentInterface[], options?: { metadatas?: Array<Record<string, any>> | null, ids?: string[] }): Promise<string[] | void> {
```
* Use Document instead of the interface (no change needed)
* Only support list as ids as options. No need to support metadata not in the document object.
```suggestion
  async addVectors(vectors: number[][], documents: Document[], options?: { ids?: string[] }): Promise<string[] | void> {
```

Itzaprado replied:
Done

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -178,15 +180,86 @@ class PostgresVectorStore extends VectorStore {
   }
 
+  async addVectors(vectors: number[][], documents: DocumentInterface[], options?: { metadatas?: Array<Record<string, any>> | null, ids?: string[] }): Promise<string[] | void> {
+    if (options && (!options.ids || options.ids.length === 0)) {
```
The Id can also come from the Document. 
* If options.ids is used - check that it is the same length as the documents. Use this id over document.id
* next check if document.id exist, if yes use that id if not generate a uuid

Itzaprado replied:
Done

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -178,15 +180,86 @@ class PostgresVectorStore extends VectorStore {
+
+    // Insert embeddings
+    for (const [id, content, embedding, metadata] of tuples) {
+      const metadataColNames = this.metadataColumns.length > 0 ? this.metadataColumns.join(",") : "";
```
The columns also need to be in quotes

Itzaprado replied:
Done

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -178,15 +180,86 @@ class PostgresVectorStore extends VectorStore {
+    for (const [id, content, embedding, metadata] of tuples) {
+      const metadataColNames = this.metadataColumns.length > 0 ? this.metadataColumns.join(",") : "";
+
+      let stmt = `INSERT INTO "${this.schemaName}"."${this.tableName}"(${this.idColumn}, ${this.contentColumn}, ${this.embeddingColumn}, ${metadataColNames}`
```
Each column name should also be in quotes

Itzaprado replied:
Done

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -178,15 +180,86 @@ class PostgresVectorStore extends VectorStore {
+    const tuples = customZip(options?.ids, documents, vectors, options?.metadatas);
+
+    // Insert embeddings
+    for (const [id, content, embedding, metadata] of tuples) {
```
nit: can we use document instead of content to make it a little more clear

Itzaprado replied:
Done

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -178,15 +180,86 @@ class PostgresVectorStore extends VectorStore {
+      let stmt = `INSERT INTO "${this.schemaName}"."${this.tableName}"(${this.idColumn}, ${this.contentColumn}, ${this.embeddingColumn}, ${metadataColNames}`
+      let values = {
+        id: id,
+        content: content.pageContent,
```
To clarify the above comment, we can keep content here but refer to the document variable for the pageContent
```suggestion
        content: document.pageContent,
```

Itzaprado replied:
Done

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -178,15 +180,86 @@ class PostgresVectorStore extends VectorStore {
+      for (const metadataColumn of this.metadataColumns) {
+        if (metadata.hasOwnProperty(metadataColumn)) {
+          valuesStmt += `, :${metadataColumn}`
+          values[metadataColumn as keyof typeof values] = metadata[metadataColumn]
```
Do we need "as keyof typeof"

Itzaprado replied:
No, I don't think so, I removed it and added a type { [key: string]: any } to values ​​so I wouldn't have an issue with this assignment.

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -178,15 +180,86 @@ class PostgresVectorStore extends VectorStore {
+   * @returns A promise resolving to an array of document IDs or void, based on implementation.
+   * @abstract
+   */
+  async addDocuments(documents: DocumentInterface[], options?: { [x: string]: any; }): Promise<string[] | void> {
```
```suggestion
  async addDocuments(documents: Document[], options?: { ids?: string[] }): Promise<string[] | void> {
```

Itzaprado replied:
Done

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -178,15 +180,86 @@ class PostgresVectorStore extends VectorStore {
+
+    for (const doc of documents) {
+      texts.push(doc.pageContent)
+      metadatas.push(doc.metadata)
```
Keep metadata in document

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -178,15 +180,86 @@ class PostgresVectorStore extends VectorStore {
+      metadatas.push(doc.metadata)
+    }
+
+    const ids = options?.hasOwnProperty('ids') ? options.ids : [];
+    const embeddings = await this.embeddings.embedDocuments(texts);
+    const results = await this.addVectors(embeddings, documents, { metadatas, ids });
```
You can just pass options through the method
```suggestion
    const embeddings = await this.embeddings.embedDocuments(texts);
    const results = await this.addVectors(embeddings, documents, options);
```

Itzaprado replied:
Done

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -178,15 +180,86 @@ class PostgresVectorStore extends VectorStore {
+      options.metadatas = Array.from(documents).map(() => new Object());
+    }
+
+    const tuples = customZip(options?.ids, documents, vectors, options?.metadatas);
```
let's check first that documents and vectors are the same array length

Itzaprado replied:
Done

Itzaprado committed:
refactor: solve comments
[ ] change options properties on addVectors and addDocuments
[ ] add if statements for ids, and documents & vector lengths

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -180,55 +180,68 @@ class PostgresVectorStore extends VectorStore {
 
-    if (options && !options.metadatas) {
-      options.metadatas = Array.from(documents).map(() => new Object());
+    if (options && options.ids && options.ids.length === documents.length) {
```
Can you throw an error if ids, vectors, and document lengths don't match like 
throw new Error(
        "The number of ids must match the number of vectors provided."
      );

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -180,55 +180,68 @@ class PostgresVectorStore extends VectorStore {
+      metadatas.push(document.metadata)
+    });
+
+    if (documents.length === vectors.length) {
```
Throw error above not the same

Itzaprado committed:
refactor: add validation for ids, vectors and documents length

Itzaprado committed:
refactor: remove unnecessary else
